File: No7_reverse.cpp
#include <math.h>

class Solution {
public:
    // 主要是防止溢出
    int reverse(int x) {
        int cur_num = 0;
        // 判断是不是-2^31, 否则会出现溢出
        if(abs(x) == pow(2, 31)){
            return 0;
        }else{
            cur_num = abs(x);
        }
        // 使用long型，避免dst_num溢出
        long dst_num = 0;
        int single = 0;
        while(cur_num > 0){
            single = cur_num % 10;
            dst_num = dst_num * 10 + single;
            cur_num = cur_num / 10;
        }
        if(x<0){
            dst_num = -dst_num;
        }
        long threshold = pow(2, 31);
        if(dst_num < -threshold || dst_num >= threshold){
            dst_num = 0;
        }
        return dst_num;
    }
};

File: No7_reverse.py
#!/usr/bin/env python
# -*- coding:utf-8 -*- 
# 重点在于判断是否溢出， 32位的范围 [-2^31, 2^31)


class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        # 判断是否为负
        is_neg = True if x < 0 else False
        num_str = str(-x) if is_neg else str(x)
        # 赋值给新的字符串
        new_num_str = ''
        for i in range(len(num_str)-1, -1, -1):
            new_num_str += num_str[i]
        print(new_num_str)
        # 判断是否会溢出
        threshold = pow(2,31)  # 2147483648
        if is_neg and int(new_num_str) <= threshold:
            new_num = -int(new_num_str)
        elif not is_neg and int(new_num_str) < threshold:
            new_num = int(new_num_str)
        else:
            new_num = 0
        return new_num
